base_layer/common_types/src/tari_address/single_address.rs

```diff
@@ -0,0 +1,378 @@
+use tari_crypto::tari_utilities::ByteArray;
+use tari_utilities::hex::{from_hex, Hex};
+
+use crate::{
+    dammsum::{compute_checksum, validate_checksum},
+    emoji::{EMOJI, REVERSE_EMOJI},
+    tari_address::{TariAddressError, TariAddressFeatures, INTERNAL_SINGLE_SIZE},
+    types::PublicKey,
+};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
+pub struct SingleAddress {
+    network: Network,
+    features: TariAddressFeatures,
+    public_spend_key: PublicKey,
+}
+
+impl SingleAddress {
+    /// Creates a new Tari Address from the provided public keys, network and features
+    pub fn new(spend_key: PublicKey, network: Network, features: TariAddressFeatures) -> SingleAddress {
+        Self {
+            network,
+            features,
+            public_spend_key: spend_key,
+        }
+    }
+
+    /// Creates a new Tari Address from the provided public keys and network while using the default features
+    pub fn new_with_default_features(spend_key: PublicKey, network: Network) -> SingleAddress {
+        Self {
+            network,
+            features: TariAddressFeatures::default(),
+            public_spend_key: spend_key,
+        }
+    }
+
+    /// helper function to convert emojis to u8
+    pub fn emoji_to_bytes(emoji: &str) -> Result<Vec<u8>, TariAddressError> {
+        // The string must be the correct size, including the checksum
+        if emoji.chars().count() != INTERNAL_SINGLE_SIZE {
+            return Err(TariAddressError::InvalidSize);
+        }
+
+        // Convert the emoji string to a byte array
+        let mut bytes = Vec::<u8>::with_capacity(INTERNAL_SINGLE_SIZE);
+        for c in emoji.chars() {
+            if let Some(i) = REVERSE_EMOJI.get(&c) {
+                bytes.push(*i);
+            } else {
+                return Err(TariAddressError::InvalidEmoji);
+            }
+        }
+        Ok(bytes)
+    }
+
+    /// Construct an TariAddress from an emoji string with checksum trying to calculate the network
+    pub fn from_emoji_string(emoji: &str) -> Result<Self, TariAddressError> {
+        let bytes = Self::emoji_to_bytes(emoji)?;
+
+        Self::from_bytes(&bytes)
+    }
+
+    /// Gets the network from the Tari Address
+    pub fn network(&self) -> Network {
+        self.network
+    }
+
+    /// Gets the features from the Tari Address
+    pub fn features(&self) -> TariAddressFeatures {
+        self.features
+    }
+
+    /// Convert Tari Address to an emoji string with checksum
+    pub fn to_emoji_string(&self) -> String {
+        // Convert the public key to bytes and compute the checksum
+        let bytes = self.to_bytes();
+        bytes.iter().map(|b| EMOJI[*b as usize]).collect::<String>()
+    }
+
+    /// Return the public spend key of an Tari Address
+    pub fn public_spend_key(&self) -> &PublicKey {
+        &self.public_spend_key
+    }
+
+    /// Construct Tari Address from bytes and try to calculate the network
```

Comment:
We calculate the entire address
```suggestion
    /// Construct Tari Address from bytes
```